refactor: update renderQuery to Suspense + Relay Hooks #5297

JimmyLv opened the pull request:
Create the PR early to make sure my first refactoring case is on the correct road, welcome code review earlier!

There are 27 remaining places that use `<QueryRender`, and these are my TODO list, which I can use in the meantime to familiarize myself with Parabol's user functionalities.

- [x] [self-check Code Review list](https://github.com/ParabolInc/parabol/blob/master/docs/codeReview.md)
- [x] ~~add basic Playwright functional tests to reduce manual test when refactoring~~ currently, there are no authenticated test cases, which need more effort (out of scope) via [packages/integration-tests/README.md](https://github.com/ParabolInc/parabol/blob/master/packages/integration-tests/README.md)
- [ ] ~~should refactoring the `createFragmentContainer` into `useFragment` in this PR? via [Realy Hooks](https://relay.dev/docs/migration-and-compatibility/relay-hooks-and-legacy-container-apis/#fragment-container--usefragment)~~
   - [ ] ~~also includes `createPaginationContainer` and `createRefetchContainer`~~
- [x] extract some common pattern of components, e.g. `<DefaultSuspense />` with fallback `Loader` as props, and `useDefaultPreloadedQuery` -> `renderLoader(options)`
- [x] check all loading spinner and related data-loading user experiences

functional use cases:

- [x] InvitationRoot.tsx
- [x] InvoiceRoot.tsx
- [x] JiraScopingSearchFilterMenuRoot.tsx
- [x] JiraScopingSearchResultsRoot.tsx
- [x] MassInvitationTokenLinkRoot.tsx
- [x] MeetingCardOptionsMenuRoot.tsx
- [x] MeetingSummaryEmailRootSSR.tsx
- [x] NewMeetingRoot.tsx
- [x] NewMeetingSummaryRoot.tsx
- [x] OrganizationRoot.tsx
- [x] OrganizationsRoot.tsx
- [x] OrgBillingRoot.tsx
- [x] OrgMembersRoot.tsx
- [x] ParabolScopingSearchResultsRoot.tsx
- [x] PokerTemplateListOrgRoot.tsx
- [x] PokerTemplateListPublicRoot.tsx
- [x] ReflectTemplateListOrgRoot.tsx
- [x] ReflectTemplateListPublicRoot.tsx
- [x] SuggestMentionableUsersRoot.tsx
- [x] TaskFooterIntegrateMenuRoot.tsx
- [x] TaskFooterTeamAssigneeMenuRoot.tsx
- [x] TaskFooterUserAssigneeMenuRoot.tsx
- [x] TeamIntegrationsRoot.tsx
- [x] TeamInvitationRoot.tsx
- [x] TeamSettingsRoot.tsx
- [x] UnpaidTeamModalRoot.tsx
- [x] UserProfileRoot.tsx
- [x] ViewerNotOnTeamRoot.tsx

Dschoordsch commented on packages/client/components/InvitationLink.tsx:
```diff
@@ -1,16 +1,28 @@
+  const {queryRef} = props
+  const data = usePreloadedQuery<InvitationLinkRootQuery>(
+    graphql`
+      query InvitationLinkRootQuery($token: ID!) {
```
-1 The query should be renamed to the place where it lives, so in this case
```
query InvitationLinkQuery($token: ID!) {
```

JimmyLv replied:
Since it's not on the Root page that needs to be modified, this should just be a naming convention issue, right? Is there a similar specification document like this?

Dschoordsch commented on packages/client/components/InvitationLink.tsx:
```diff
@@ -1,16 +1,28 @@
+      }
+    `,
+    queryRef,
+    {UNSTABLE_renderPolicy: 'full'}
```
Fun fact, I googled what that means, this was my first result: https://www.parabol.co/blog/upgrade-react-to-graphql-relay-hooks/ Kudos @mattkrick 
TL;DR: we need this until we get rid of `createFragmentContainer`

Dschoordsch commented on packages/client/components/TeamInvitation.tsx:
```diff
@@ -1,17 +1,31 @@
+  const {queryRef} = props
+  const data = usePreloadedQuery<TeamInvitationRootQuery>(
+    graphql`
+      query TeamInvitationRootQuery($token: ID!) {
```
-1 see above, please rename, same in the files following

JimmyLv replied:
sure, thanks Georg @Dschoordsch

Dschoordsch reviewed (changes requested):
Looks good so far!

JimmyLv commented:
@Dschoordsch Hey Georg! I have refactored all the component files related to `QueryRenderer` and removed the`renderQuery`, could you please help to review again? 🤩
I have checked the availability of the functionalities _before and after_ (when doing each Git commit), but can't guarantee that all the refactoring is exactly as expected by the product.

This is my first PR and any feedback/suggestions are welcome, thank you so much! ✌️

JimmyLv commented on packages/client/components/NewMeeting.tsx:
```diff
@@ -118,12 +116,38 @@ const createMeetingOrder = ({standups}: {standups: boolean}) => {
   const newMeetingOrder = useMemo(() => createMeetingOrder(featureFlags), [featureFlags])
 
-  useStoreQueryRetry(retry)
+  // useStoreQueryRetry(retry)
```
actually, I'm not sure if the `retry` is still needed here, since the `refetch` is already set inside the `useQueryLoaderNow`

JimmyLv commented on packages/client/modules/teamDashboard/containers/TeamIntegrationsRoot/TeamIntegrationsRoot.tsx:
```diff
@@ -1,33 +1,25 @@
-  const atmosphere = useAtmosphere()
+  const queryRef = useQueryLoaderNow<ProviderListQuery>(providerListQuery, {teamId})
+  const [, loadQuery] = useQueryLoader<ProviderListQuery>(providerListQuery)
+  const retry = () => {
```
this `retry` also need some attention 🤔

Dschoordsch replied:
+1 should work for now. A nicer solution would be to have a version of `useQueryLoaderNow` which exposes `retry()`
```
export const useQueryLoaderNowWithRetry = <TQuery extends OperationType>(
   ...
   return {queryRef, retry: refreshQuery}
}
const useQueryLoaderNow = ...
  const {queryRef} = useQueryLoaderNowWithRetry(...)
  return queryRef
}
```

JimmyLv commented on packages/client/modules/userDashboard/containers/OrgMembers/OrgMembersRoot.tsx:
```diff
@@ -20,16 +21,26 @@ interface Props {
-      fetchPolicy={'store-or-network' as any}
-      render={renderQuery(OrgMembers, {size: LoaderSize.PANEL})}
-    />
+    <Suspense fallback={''}>{queryRef && <OrgMembersContainer queryRef={queryRef} />}</Suspense>
```
for these Suspense `fallback` here, not quite sure if I need to add back the `Loader.Size` Loader component at this level

Dschoordsch replied:
-1 I think we should have a `MockList` fallback here

JimmyLv commented on packages/client/modules/userDashboard/containers/OrgBilling/OrgBillingRoot.tsx:
```diff
@@ -20,18 +21,31 @@ interface Props {
+  organization: OrgBillingRoot_organization
+}
+
+function OrgBillingContainer(props: OrgBillingContainerProps) {
```
This pattern would be simpler without involving modifications to the child component, e.g. `OrgBilling` (especially if it involves the `createPaginationContainer` and `createRefetchContainer` HOC components)

Because it uses `usePreloadedQuery` to request data, I call it the `Container` of component (compare to `Wrapper`)

mattkrick replied:
makes sense, I'll backlog that work in https://github.com/ParabolInc/parabol/issues/6431

JimmyLv commented on packages/client/components/JiraScopingSearchResultsRoot.tsx:
```diff
@@ -1,55 +1,38 @@
-        return <JiraScopingSearchResults viewer={viewer} meeting={meeting} />
-      }}
-    />
+    <ErrorBoundary fallback={(error) => <ErrorComponent error={error} eventId={''} />}>
```
Not quite sure if I need to pass the `eventId` in from the `fallback()` parameter of the `ErrorBoundary`

JimmyLv replied:
cool, I just saw the `render()` method of `ErrorComponent`, so by default, we don't need the error fallback.

Dschoordsch replied:
Yes, that's also something I learned while doing this review

JimmyLv commented on packages/client/components/JiraScopingSearchResults.tsx:
```diff
@@ -36,6 +83,7 @@ const JiraScopingSearchResults = (props: Props) => {
+  // TODO: after add Suspense fallback in Root, should be able to remove this?
   // even though it's a little herky jerky, we need to give the user feedback that a search is pending
   // TODO fix flicker after viewer is present but edges isn't set
   if (!edges) {
```
`<MockScopingList />` here is a duplicate with the `fallback` of Suspense at the upper level

JimmyLv replied:
okay, so I also remove the related comments here.

Dschoordsch commented on packages/client/modules/outcomeCard/components/OutcomeCardAssignMenu/TaskFooterTeamAssigneeMenu.tsx:
```diff
@@ -170,19 +194,6 @@ const TaskFooterTeamAssigneeMenu = (props: Props) => {
 }
 
 export default createFragmentContainer(TaskFooterTeamAssigneeMenu, {
```
-1 please convert it fully away from `createFragmentContainer`

Dschoordsch commented on packages/client/modules/outcomeCard/components/TaskFooterTeamAssigneeMenuRoot.tsx:
```diff
@@ -24,16 +15,17 @@ interface Props {
-      fetchPolicy={'store-or-network' as any}
-      render={renderQuery(TaskFooterTeamAssigneeMenu, {props: {menuProps, task}})}
-    />
+    <Suspense fallback={''}>
```
-1 `renderQuery` had a loading component, probably should be `MockList`

Dschoordsch commented on packages/client/modules/outcomeCard/components/TaskFooterUserAssigneeMenuRoot.tsx:
```diff
@@ -28,15 +23,21 @@ const TaskFooterUserAssigneeMenuRoot = (props: Props) => {
-      render={renderQuery(TaskFooterUserAssigneeMenu, {props: {area, menuProps, task}})}
-      fetchPolicy={'store-or-network' as any}
-    />
+    <Suspense fallback={''}>
```
-1 `renderQuery` had a fallback

Dschoordsch commented on packages/client/modules/summary/components/NewMeetingSummaryRoot.tsx:
```diff
@@ -1,37 +1,23 @@
-      render={renderQuery(NewMeetingSummary, {props: {urlAction}, size: LoaderSize.WHOLE_PAGE})}
-      fetchPolicy={'store-or-network' as any}
-    />
+    <Suspense fallback={''}>
```
-1 `renderQuery` had a fallback

Dschoordsch commented on packages/client/modules/teamDashboard/components/TeamSettingsRoot.tsx:
```diff
@@ -1,34 +1,17 @@
-    />
-  )
+  const queryRef = useQueryLoaderNow<TeamSettingsQuery>(teamSettingsQuery, {teamId})
+  return <Suspense fallback={''}>{queryRef && <TeamSettings queryRef={queryRef} />}</Suspense>
```
-1 `renderQuery` had a fallback

Dschoordsch commented on packages/client/modules/teamDashboard/containers/UnpaidTeamModal/UnpaidTeamModalRoot.tsx:
```diff
@@ -1,36 +1,19 @@
-  )
+  const {teamId} = props
+  const queryRef = useQueryLoaderNow<UnpaidTeamModalQuery>(unpaidTeamModalQuery, {teamId})
+  return <Suspense fallback={''}>{queryRef && <UnpaidTeamModal queryRef={queryRef} />}</Suspense>
```
-1 `renderQuery` had a fallback

Dschoordsch commented on packages/client/utils/relay/renderFallback.tsx:
```diff
@@ -0,0 +1,37 @@
+import React, {ComponentType, ReactNode} from 'react'
```
-1 the file name is not ideal, `renderLoader` would be more fitting

Dschoordsch commented on packages/client/modules/userDashboard/containers/Organization/OrganizationRoot.tsx:
```diff
@@ -21,16 +13,8 @@ const OrganizationRoot = (props: Props) => {
-    />
-  )
+  const queryRef = useQueryLoaderNow<OrganizationQuery>(organizationQuery, {orgId})
+  return <Suspense fallback={''}>{queryRef && <Organization queryRef={queryRef} />}</Suspense>
```
-1 `renderQuery` had a fallback

Dschoordsch commented on packages/client/modules/userDashboard/containers/OrgBilling/OrgBillingRoot.tsx:
```diff
@@ -20,18 +21,31 @@ interface Props {
+  return <OrgBilling viewer={viewer} organization={organization} />
+}
+
 export default createFragmentContainer(OrgBillingRoot, {
```
-1 If you touch this already, please convert the `createFragmentContainer` call to make it more readable overall

Dschoordsch commented on packages/client/modules/userDashboard/components/UserSettingsForm/UserSettingsForm.tsx:
```diff
@@ -0,0 +1,129 @@
+import useAtmosphere from '../../../../hooks/useAtmosphere'
```
+1 It probably made sense to move the logic out, but it makes review harder if unrelated changes are intermingled. No need to revert this, but just a note for the future.

Dschoordsch replied:
If there are dependencies, one PR can target the feature branch of another PR instead of master. Once the first branch is merged, GitHub will change the target branch of the second one to master.

JimmyLv replied:
cool, make sense! thanks @Dschoordsch

Dschoordsch reviewed (changes requested):
I'm not sure if you had a reason to remove the render fallback in many cases?

JimmyLv commented on packages/client/components/AvatarInput.tsx:
```diff
@@ -11,7 +11,7 @@ const Control = styled('div')({
   justifyContent: 'center',
   overflow: 'hidden',
-  padding: '.5rem',
+  padding: '.5rem 1rem 1rem',
```
since it is too trivial so I just take it easy to fix 😁

https://github.com/ParabolInc/parabol/issues/6169#issuecomment-1075471314

JimmyLv commented on packages/client/components/ParabolScopingSearchResultsRoot.tsx:
```diff
@@ -29,31 +33,42 @@ interface Props {
+  return <ParabolScopingSearchResults viewer={viewer} meeting={meeting} />
+}
+
 export default createFragmentContainer(ParabolScopingSearchResultsRoot, {
```
`...Container` here, because I'm not going to deal with the `createPaginationContainer` inside the `ParabolScopingSearchResult` component yet, maybe we can consider refactoring it to `usePaginationFragment` in #6283

JimmyLv replied:
but I did the refactoring to replace createFragmentContainer with `useFragment` for ParabolScopingSearchResultsRoot

JimmyLv commented on packages/client/hooks/useQueryLoaderNow.tsx:
```diff
@@ -6,14 +6,17 @@ import {
   fetchPolicy?: PreloadFetchPolicy,
-  preventSuspense?: boolean
+  preventSuspense?: boolean,
+  relayEnvironment?: Environment
```
this one for MeetingSummaryEmailRootSSR https://github.com/ParabolInc/parabol/commit/8b474c9a45e9ca2b450a990b2416ee091f69e44e

JimmyLv commented on packages/client/modules/email/components/MeetingSummaryEmailRootSSR.tsx:
```diff
@@ -34,36 +37,57 @@ export const meetingSummaryUrlParams = {
+    environment
+  )
+  return (
+    <Suspense fallback={''}>
```
What I'm thinking of is removing the `Suspense`, but not quite sure if the environment specified here works on SSR.  @mattkrick have any ideas?

Another thing is: about the feature of sending emails, how do I go about verifying it in the PR phase?

Dschoordsch replied:
There is a dummy mail provider which just prints the email's plain text to stdout. I guess it would be easy to modify it to print the html instead.

JimmyLv commented:
Hi @Dschoordsch I have solved most of the feedback, could you please help me to re-review it? 🙋‍♂️

JimmyLv commented:
@Dschoordsch Hey Georg, how about other feedback? and should I move this PR into the maintainer review column?

Dschoordsch commented:
@JimmyLv sorry it takes so long. I got stuck with the review on the 2 places where you created a container as a double wrapper. I'm not happy getting it merged like that, but didn't had the time to look into a neater solution as I'm also lacking experience with the relay hooks, so I don't know how these interact with pagination. I will set aside some time today to check.

Dschoordsch reviewed (approved):
I think having suspense in the email rendering server side does not do any harm and at least is consistent with the rest of the components [packages/client/modules/email/components/MeetingSummaryEmailRootSSR.tsx](https://github.com/ParabolInc/parabol/pull/6251/files/ac2c01a7c5b9c18949be374553e4d5a55846d1b7#diff-b83de9a3cc74b538222bad9310572bd10997c035ac8e10beb5aa178fc6d10fdd). 
I couldn't find a nicer solution for
* [packages/client/components/ParabolScopingSearchResultsRoot.tsx](https://github.com/ParabolInc/parabol/pull/6251/files/789d0c0f7cfaff2541d642ff8bd82a1065a8f7bd#diff-73ba6307f98e0ca59f8244045cacc31b699818b7ff2ce086b5f4712983f287f3)
* [packages/client/modules/userDashboard/containers/OrgBilling/OrgBillingRoot.tsx](https://github.com/ParabolInc/parabol/pull/6251/files/ac2c01a7c5b9c18949be374553e4d5a55846d1b7#diff-19053e2335214fa597bca5c90280d4f99f71cc21cde3d98f8bb81b70b4cb283f)

without refactoring to `usePaginationFragment` which requires even more refactoring.

Let's see what @mattkrick has to say about it.

JimmyLv commented:
@Dschoordsch Thanks for the detailed review! 

about the `usePaginationFragment`, I think it would be better if we consider refactoring these Legacy Container APIs to Relay Hooks together (maybe in #6283). I found several areas that needed to be refactored, including `createPaginationContainer` and `createRefetchContainer`.

ref: [Relay Hooks and Legacy Container APIs | Relay](https://relay.dev/docs/migration-and-compatibility/relay-hooks-and-legacy-container-apis/#refetch-container--userefetchablefragment)

mattkrick commented on packages/client/modules/email/components/MeetingSummaryEmailRootSSR.tsx:
```diff
@@ -34,36 +37,55 @@ export const meetingSummaryUrlParams = {
 
 const MeetingSummaryEmailRootSSR = (props: Props) => {
   const {appOrigin, environment, meetingId} = props
+  const queryRef = useQueryLoaderNow<MeetingSummaryEmailRootSSRQuery>(
```
-1 i believe we'll have to change this because now we're importing more than just a type, we're importing an object.

This is problematic because to build `meetingSummaryEmailRootSsrQuery`, relay-compiler needs a graphql schema & to build a graphql schema, graphql must traverse the /graphql/public dir, which eventually depends on this file. This creates a circular dependency.

I'd suggest using a lazy-load pattern. See https://github.com/ParabolInc/parabol/blob/0cb19afc1043501a55a990e3a838c1792840e1bb/packages/client/components/DemoMeetingRoot.tsx#L30

JimmyLv replied:
make sense! I updated to use `useLazyLoadQuery` to load data directly and avoid circular dependency

mattkrick commented on packages/client/modules/meeting/components/PokerTemplateListOrgRoot.tsx:
```diff
@@ -21,16 +13,14 @@ interface Props {
 const PokerTemplateListOrgRoot = (props: Props) => {
   const {isActive, teamId} = props
-  const atmosphere = useAtmosphere()
+  const queryRef = useQueryLoaderNow<PokerTemplateListOrgQuery>(pokerTemplateListOrgQuery, {teamId})
```
-1 we don't want to trigger a network request until `isActive` is true. easiest fix is to probably move the `isActive` logic to the parent.

TEST
- [ ] open dev tools -> network -> websocket & monitor traffic
- [ ] open the poker meeting settings
- [ ] this query should _not_ get fetched until the organization tab is clicked

JimmyLv replied:
good point! This is the difference between component-based usage and React hook, and I've updated it!

mattkrick commented on packages/client/modules/meeting/components/PokerTemplateListPublicRoot.tsx:
```diff
@@ -21,16 +13,15 @@ interface Props {
 const PokerTemplateListPublicRoot = (props: Props) => {
   const {isActive, teamId} = props
-  const atmosphere = useAtmosphere()
+  const queryRef = useQueryLoaderNow<PokerTemplateListPublicQuery>(pokerTemplateListPublicQuery, {
```
-1 same as above, network request only on isActive

mattkrick commented on packages/client/modules/meeting/components/ReflectTemplateListOrgRoot.tsx:
```diff
@@ -21,16 +13,14 @@ interface Props {
 const ReflectTemplateListOrgRoot = (props: Props) => {
   const {isActive, teamId} = props
-  const atmosphere = useAtmosphere()
+  const queryRef = useQueryLoaderNow<ReflectTemplateListOrgQuery>(reflectTemplateListOrgQuery, {
```
-1 same as above, network request only on isActive

mattkrick commented on packages/client/modules/meeting/components/ReflectTemplateListPublicRoot.tsx:
```diff
@@ -21,16 +13,17 @@ interface Props {
 const ReflectTemplateListPublicRoot = (props: Props) => {
   const {isActive, teamId} = props
-  const atmosphere = useAtmosphere()
+  const queryRef = useQueryLoaderNow<ReflectTemplateListPublicQuery>(
```
-1 same as above, network request only on isActive

mattkrick commented on packages/client/modules/outcomeCard/components/TaskFooterTeamAssigneeMenuRoot.tsx:
```diff
@@ -24,16 +16,17 @@ interface Props {
-      fetchPolicy={'store-or-network' as any}
-      render={renderQuery(TaskFooterTeamAssigneeMenu, {props: {menuProps, task}})}
-    />
+    <Suspense fallback={MockFieldList}>
```
+2 nice touch adding a fallback here!

mattkrick commented on packages/client/modules/summary/components/NewMeetingSummaryRoot.tsx:
```diff
@@ -1,37 +1,25 @@
 }
 
-export default withAtmosphere(withRouter(NewMeetingSummaryRoot))
+export default withRouter(NewMeetingSummaryRoot)
```
+1 `withRouter` is deprecated, you could replace it with `useRouter` if you like, although if this PR is too big, I understand

mattkrick reviewed (changes requested):
nice work! couple small things, let me know if they give you trouble fixing

JimmyLv commented:
@mattkrick Thanks for the review! Victory is at hand 🤩
I did update with feedback, including `useLazyLoadQuery` and network requests for Poker and Reflect TemplateList, etc. 
Then I did a merge with the master and I think `git precommit` hook did some automatic formatting while handling the conflict, hopefully it will not be too disturbing.

mattkrick commented:
LGTM! no worries about the formatting, we'll run https://github.com/ParabolInc/parabol/issues/6314 in the next sprint & get the whole repo formatted correctly so we don't encounter those things anymore
